pkg/go/graph/weighted_graph.go

```diff
@@ -0,0 +1,399 @@
+package graph
+
+import (
+	"fmt"
+	"math"
+	"strings"
+)
+
+const Infinite = math.MaxInt32
+
+type WeightedAuthorizationModelGraph struct {
+	nodes map[string]*WeightedAuthorizationModelNode
+	edges map[string][]*WeightedAuthorizationModelEdge
+}
+
+// NewGraph creates a new Graph.
+func NewWeightedAuthorizationModelGraph() *WeightedAuthorizationModelGraph {
+	return &WeightedAuthorizationModelGraph{
+		nodes: make(map[string]*WeightedAuthorizationModelNode),
+		edges: make(map[string][]*WeightedAuthorizationModelEdge),
+	}
+}
+
+// AddNode adds a node to the graph with optional operationType and weight.
+func (wg *WeightedAuthorizationModelGraph) AddNode(uniqueLabel, label string, nodeType NodeType) {
+	wg.nodes[uniqueLabel] = &WeightedAuthorizationModelNode{uniqueLabel: uniqueLabel, label: label, nodeType: nodeType}
+}
+
+func (wg *WeightedAuthorizationModelGraph) AddEdge(fromID, toID string, edgeType EdgeType, condition string) {
+	fromNode := wg.nodes[fromID]
+	toNode := wg.nodes[toID]
+	edge := &WeightedAuthorizationModelEdge{from: fromNode, to: toNode, edgeType: edgeType, conditionedOn: condition}
+	wg.edges[fromID] = append(wg.edges[fromID], edge)
+}
+
+func (wg *WeightedAuthorizationModelGraph) AssignWeights() error {
+	visited := make(map[string]bool)
+	ancestorPath := make([]*WeightedAuthorizationModelEdge, 0)
+	tupleCycleDependencies := make(map[string][]*WeightedAuthorizationModelEdge)
+
+	for node := range wg.nodes {
+		if visited[node] {
+			continue
+		}
+
+		tupleCyles, err := wg.calculateNodeWeight(node, visited, ancestorPath, tupleCycleDependencies)
+		if err != nil {
+			return err
+		}
+		if len(tupleCyles) > 0 {
+			for _, tupleCycle := range tupleCyles {
+				fmt.Println("Tuple cycle is ", tupleCycle)
+			}
+			return fmt.Errorf("%d tuple cycles found without resolution", len(tupleCyles))
+		}
+	}
+	return nil
+}
+
+func (wg *WeightedAuthorizationModelGraph) calculateNodeWeight(nodeID string, visited map[string]bool, ancestorPath []*WeightedAuthorizationModelEdge, tupleCycleDependencies map[string][]*WeightedAuthorizationModelEdge) ([]string, error) {
+	tupleCycles := make([]string, 0)
+
+	if visited[nodeID] {
+		return nil, nil
+	}
+
+	if wg.nodes[nodeID].nodeType == SpecificType || wg.nodes[nodeID].nodeType == SpecificTypeWildcard {
+		return nil, nil
+	}
+
+	visited[nodeID] = true
+
+	for _, edge := range wg.edges[nodeID] {
+		if len(edge.weights) != 0 {
+			continue
+		}
+		if edge.to.nodeType == SpecificType || edge.to.nodeType == SpecificTypeWildcard {
+			edge.weights = make(map[string]int)
+			uniqueLabel := edge.to.uniqueLabel
+			if edge.to.nodeType == SpecificTypeWildcard {
+				uniqueLabel = uniqueLabel[:len(uniqueLabel)-2]
+			}
+			edge.weights[uniqueLabel] = 1
+			continue
+		}
+
+		tcycle, err := wg.calculateEdgeWeight(edge, ancestorPath, visited, tupleCycleDependencies)
+		if err != nil {
+			tupleCycles = append(tupleCycles, tcycle...)
```

Comment:
Why, if we aren't going to be inspecting it?

Reply:
for correctness on the step.